src/sentry/auth/helper.py

```diff
@@ -226,12 +227,57 @@ def _get_login_redirect(self, subdomain: str | None) -> str:
             login_redirect_url = absolute_uri(login_redirect_url, url_prefix=url_prefix)
         return login_redirect_url
 
-    def _handle_new_membership(self, auth_identity: RpcAuthIdentity) -> RpcOrganizationMember:
-        user, om = auth_service.handle_new_membership(
+    def _handle_membership(
+        self,
+        request: Request,
+        organization: RpcOrganization,
+        auth_identity: AuthIdentity,
+    ) -> Tuple[User, RpcOrganizationMember]:
+        # TODO: Might be able to keep hold of the RpcUser object whose ID was
+        #  originally passed to construct the RpcAuthIdentity object
+        user = User.objects.get(id=auth_identity.user_id)
+
+        # If the user is either currently *pending* invite acceptance (as indicated
+        # from the invite token and member id in the session) OR an existing invite exists on this
+        # organization for the email provided by the identity provider.
+        invite_helper = ApiInviteHelper.from_session_or_email(
+            request=request, organization_id=organization.id, email=user.email
+        )
+
+        # If we are able to accept an existing invite for the user for this
+        # organization, do so, otherwise handle new membership
+        if invite_helper:
+            if invite_helper.invite_approved:
+                rpc_om = invite_helper.accept_invite(user)
+                assert rpc_om
+                return user, rpc_om
+
+            # It's possible the user has an _invite request_ that hasn't been approved yet,
+            # and is able to join the organization without an invite through the SSO flow.
+            # In that case, delete the invite request and create a new membership.
+            invite_helper.handle_invite_not_approved()
+
+        flags = RpcOrganizationMemberFlags(sso__linked=True)
+        # if the org doesn't have the ability to add members then anyone who got added
+        # this way should be disabled until the org upgrades
+        if not features.has("organizations:invite-members", organization):
+            flags.member_limit__restricted = True
+
+        # Otherwise create a new membership
+        om = organization_service.add_organization_member(
+            organization_id=organization.id,
+            default_org_role=organization.default_role,
+            role=organization.default_role,
+            user_id=user.id,
+            flags=flags,
+        )
+        return user, om
+
+    def _handle_new_membership(self, auth_identity: AuthIdentity) -> RpcOrganizationMember:
+        user, om = self._handle_membership(
```

Comment:
`om` is always non-`None` now, right? Can we delete the `if om is not None` check below?